composition/api/v1alpha1/composition_types.go

```diff
@@ -0,0 +1,217 @@
+	//InputNamespace string     `json:"inputNamespace,omitempty"`
+	//Sinc      Sinc       `json:"sinc,omitempty"`
+
+	Description string `json:"description,omitempty"`
+
+	// TODO (barney -s) rename to FacadeAPIGroup,facadeAPIGroup
+
+	// Use existing KRM API
+	InputAPIGroup string `json:"inputAPIGroup,omitempty"`
+
+	//+kubebuilder:validation:MinItems=1
+	Expanders []Expander `json:"expanders"`
+	// Namespace mode indicates how compositions set the namespace of the objects from expanders.
+	// ""|inherit implies inherit the facade api's namespace. Only namespaced objects are allowed.
+	// explicit     implies the objects in the template must have the namespace set.
+	// +kubebuilder:validation:Enum=inherit;explicit
+	NamespaceMode NamespaceMode `json:"namespaceMode,omitempty"`
+
+	// Readiness
+	Readiness []ReadyOn `json:"readiness,omitempty"`
+}
+
+type ValidationStatus string
+
+const (
+	// ValidationStatusUnkown is when it is not validated
+	ValidationStatusUnknown ValidationStatus = "unknown"
+	// ValidationStatusSuccess is when valdiation succeeds
+	ValidationStatusSuccess ValidationStatus = "success"
+	// ValidationStatusFailed is when valdiation fails
+	ValidationStatusFailed ValidationStatus = "failed"
+	// ValidationStatusError is when validation was not called
+	ValidationStatusError ValidationStatus = "error"
+)
+
+// StageStatus captures the status of a stage
+type StageValidationStatus struct {
+	ValidationStatus ValidationStatus `json:"validationStatus,omitempty"`
+	Reason           string           `json:"reason,omitempty"`
+	Message          string           `json:"message,omitempty"`
+}
+
+// CompositionStatus defines the observed state of Composition
+type CompositionStatus struct {
+	Generation int64                            `json:"generation,omitempty"`
+	Conditions []metav1.Condition               `json:"conditions,omitempty"`
+	Stages     map[string]StageValidationStatus `json:"stages,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// Composition is the Schema for the compositions API
+type Composition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   CompositionSpec   `json:"spec,omitempty"`
+	Status CompositionStatus `json:"status,omitempty"`
+}
+
+//+kubebuilder:object:root=true
+
+// CompositionList contains a list of Composition
+type CompositionList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []Composition `json:"items"`
+}
+
+// Status helpers
+func (s *CompositionStatus) ClearCondition(condition ConditionType) {
+	meta.RemoveStatusCondition(&s.Conditions, string(condition))
+}
+
+// Validation
+func (s *Composition) Validate() bool {
+	// Several of these validations should br CEL rules on the composition CRD
+	// However for now they help me shape the controller.
+	s.Status.ClearCondition(ValidationFailed)
+	// Validate Expanders
+	message := ""
```

Comment:
Can't this just be a `kubebuilder:validation:Required` check?

Reply:
It can. Removed that bit. Retaining the Validation function for now. At some point we can remove it as well. Thinking of using this to check presence of expanderversion objects later on. Added to TODO.md

Reply:
Removed the whole function